Size shifting feature

YeDemirkiran opened the pull request:
- Created the Nomad object and made the Player's mesh its children. Then made the Nomad the Player's child.

- Duplicated the Nomad mesh two times, one is a small one named Pixie and the other one is the largest mesh, named Titan. Added Capsule Colliders to each of them, according to their sizes.

- Added the Nomad, Pixie and Titan actions to the Input Map. Then added Nomad, Pixie and Titan fields to the PlayerMovement scripts and OnPixie, OnTitan and OnNomad methods for setting the objects to the desired state.

- Added the jumpCounter and maxJumpCounter fields. The player only can jump if the jumpCounter is less than maxJumpCounter.

  - The jumpCounter gets incremented by 1 when the OnJump method is called.

  - It resets to 0 when the player is on the ground, so they can jump again.

  - Nomad can jump one time, Pixie can jump two times, and Titan can't jump.

YeDemirkiran committed:
Created the Nomad object as the child of the Player object

Created the Nomad object and made the Player's mesh its children. Then made the Nomad the Player's child.

YeDemirkiran committed:
Created the Pixie and Titan meshes

-Duplicated the Nomad mesh two times, one is a small one named Pixie and the other one is the largest mesh, named Titan

-Added Capsule Colliders to each of them, according to their sizes

YeDemirkiran committed:
Added the basic size shifting

-Added the Nomad, Pixie and Titan actions to the Input Map

-Added Nomad, Pixie and Titan fields to the PlayerMovement scripts and OnPixie, OnTitan and OnNomad methods for settng the objects to the desired state.

YeDemirkiran committed:
Added the Jump Counter mechanism to PlayerMovement

-Added the jumpCounter and maxJumpCounter fields. The player only can jump if the jumpCounter is less than maxJumpCounter

-The jumpCounter gets incremented by 1 when the OnJump method is called

-It resets to 0 when the player is on the ground, so they can jump again

-Nomad can jump one time, Pixie can jump two times, and Titan can't jump

YeDemirkiran committed:
Set the Nomad to be the default size at start

zamcham commented on Assets/Scripts/PlayerController.cs:
```diff
@@ -79,11 +88,87 @@ void GetGroundChecker()
+    {
+        Debug.Log("Titan");
+
+        titan.SetActive(true);
+
+        pixie.SetActive(false);
+        nomad.SetActive(false);
```
To avoid issues with colliders when we add the animations please add the `SetActive(false);` functions first before activating the new G.O

YeDemirkiran replied:
Done

zamcham commented on Assets/Scripts/PlayerController.cs:
```diff
@@ -79,11 +88,87 @@ void GetGroundChecker()
+    {
+        Debug.Log("Nomad");
+
+        nomad.SetActive(true);
+
+        pixie.SetActive(false);
+        titan.SetActive(false);
```
To avoid issues with colliders when we add the animations please add the `SetActive(false);` functions first before activating the new G.O

YeDemirkiran replied:
Done

zamcham commented on Assets/Scripts/PlayerController.cs:
```diff
@@ -79,11 +88,87 @@ void GetGroundChecker()
+    {
+        Debug.Log("Pixie");
+
+        pixie.SetActive(true);
+
+        titan.SetActive(false);
+        nomad.SetActive(false);
```
To avoid issues with colliders when we add the animations please add the `SetActive(false);` functions first before activating the new G.O

YeDemirkiran replied:
Done

zamcham commented on Assets/Scripts/PlayerController.cs:
```diff
@@ -79,11 +88,87 @@ void GetGroundChecker()
+        maxJumpCount = 2;
+    }
+
+    void GetCharacters()
+    {
+        Transform pixie = transform.Find("Pixie");
+        Transform nomad = transform.Find("Nomad");
+        Transform titan = transform.Find("Titan");
+
+        if (pixie == null)
+        {
+            Debug.LogError("Pixie mesh not found!");
+        }
+        else
+        {
+            this.pixie = pixie.gameObject;
+        }
+
+        if (nomad == null)
+        {
+            Debug.LogError("Nomad mesh not found!");
+        }
+        else
+        {
+            this.nomad = nomad.gameObject;         
+        }
+
+        if (titan == null)
+        {
+            Debug.LogError("Titan mesh not found!");
+        }
+        else
+        {
```
Great job here with the logic, just a tip, for if/else repetitive statements you can use `switch` cases instead, here's a cleaner version:

```
void GetCharacters()
{
    string[] forms = { "Pixie", "Nomad", "Titan" };

    foreach (string form in forms)
    {
        Transform currentForm = transform.Find(form);

        if (currentForm == null)
        {
            Debug.LogError($"{form} mesh not found!");
        }
        else
        {
            switch (form)
            {
                case "Pixie":
                    this.pixie = currentForm.gameObject;
                    break;
                case "Nomad":
                    this.nomad = currentForm.gameObject;
                    break;
                case "Titan":
                    this.titan = currentForm.gameObject;
                    break;
            }
        }
    }
}
```

Feel free to use the code above if you'd like

YeDemirkiran replied:
Thank you :) Done.

zamcham reviewed (changes requested):
Great job @YeDemirkiran 👍  you just need to make a few changes before I approve the request :)

YeDemirkiran committed:
Review changes & code cleanup

-Changed the order of the object activation on size-shrinking methods: First deactivation of the two other objects, then the activation of the selected size.

-Switch statements instead of If statements inside the GetCharacters method

YeDemirkiran committed:
Added the UI canvas and Stamina Bar

-Created the UI canvas, set its scale mode to 'Scale With Screen Size' and chose 1920x1080 as its reference size.

-Created an uninteractable slider to use as a stamina bar and placed it on the top left corner of the screen. Anchors and the pivot on the top left.

YeDemirkiran committed:
Created the EnergyManagement script

With 1 property and 4 private fields:

-currentEnergy: To monitor the current energy

-maxEnergy: The energy won't pass this value

-startingEnergy: The energy will start at this value at the start of the level

-pixieEnergyThreshold: The amount of energy requred to shrink in size (become Pixie)

-titanEnergyThreshold: The amount of energy requred to grow in size (become Titan)

YeDemirkiran committed:
Created a general purpose BarUI script

-The script has a private Slider field. It gets the Slider component in the Start method, and sets its max value to the serialzied private float 'maxValue'.

-It has one public method for setting its value. When this method is called, a coroutine is started internally to lerp the current value to the target value. The duration of this transition is set by serialized private field 'valueLerpDuration'.

YeDemirkiran committed:
Integrated the Energy System with the Size Shifting

-On each size change, the PlayerMovement will check in order:

--Whether if already changed to that size, or not

--If not, it checks if there is enough energy

-After these conditions are met, the new size is activated and the required amount is subtracted from the currentEnergy value of the EnergyManagement instance.

-The EnergyManagement script has a serialized BarUI field and inside the AddEnergy method, the SetValue of the BarUI is called.

YeDemirkiran committed:
Moved the new created scripts into the Scripts folder

-BarUI

-EnergyManagement

-TriggerCollision

YeDemirkiran committed:
New folders & assets

-Prefabs

-Prefabs > Collectibles

-Sprites

-Sprites > Collectibles

And a placholder orb sprite.

YeDemirkiran committed:
Scripts folder re-organized & Orb sprite settings

-Changed the placeholder orb's Pixel per Unit settings to 256 from 100

-New folders inside the Scripts folder

YeDemirkiran committed:
Created the ICollectible interface and EnergyOrb component

-ICollectible: A general purpose collectible class with an OnCollect public method, with a GameObject parameter 'collector'. Useful for powerups and collectibles. Might extend the methods for other uses.

-EnergyOrb: Implements the MonoBehaviour class and ICollectible interface. One serialized private float 'energyGain', and public method OnCollect.

It gets the energyManager component from the passed in parameter 'collector', and calls its AddEnergy method. Then, it destroys itself.

YeDemirkiran committed:
Clamped the currentEnergy field inside AddEnergy()

EnergyManagement class' AddEnergy method doesn't clamp the value. It was certain that this would cause the value to go out of the bounds. I forgot to clamp it between 0 and the maxEnergy field when first created this method. Now done.

YeDemirkiran committed:
Added the Orb pick-up

-Created the Collectible layer

-Added this layer to the EnergyOrb object

-Added an OnTriggerEnter2D method to the PlayerMovement script, and called the OnCollect method of the collided collectible

YeDemirkiran committed:
Added sound to the Energy Orb

-Created the 'Audio/SFX' folders

-Created a placeholder pick-up item sound

-The OnCollect method of the EnergyOrb will play a sound using the PlayClipAtPoint static method. This uses the 3D sound system instead of 2D, we probably need a better solution.

-Placed several Energy Orbs around the scene

YeDemirkiran committed:
Merge pull request #9 from zamcham/pickup-system

Pick-up system

YeDemirkiran committed:
Added the AddEnergyPercent on EnergyManagement

Compared to the other AddEnergy method, this method treats the float parameter as a percent, divides it by 100 and then multiplies the 'maxEnergy' by it. Then adds the value to the currentEnergy.

YeDemirkiran committed:
Converted the energy system to percentage based

-Added 3 new private floats: pixieEnergyThreshold, titanEnergyThreshold, nomadEnergyCost

-The size change methods in PlayerMovement now will check for the percentage of the currentEnergy, and subtract energy using the AddEnergyPercent method, instead of the AddEnergy method

YeDemirkiran committed:
EnergyOrb uses percentage

The orbs now use the AddEnergyPercent method.

zamcham reviewed (changes requested):
@YeDemirkiran Almost there! Looks like we start with default stamina, let's set it up to 0 to begin the game

Also, looks like we can switch ti Titan without having the bar full at 100%, we should only be able to do it if the bar if full :)

zamcham reviewed (approved):
Ready to merge @YeDemirkiran 💯
